File: examples/client/tutorials/transactions/simple_transactions.js
// Install necessary polyfills (see supported browsers) into global
import 'core-js/es6';
import 'core-js/stage/4';
import 'isomorphic-fetch';

import {
    Ed25519Keypair,
    makeCreateTransaction,
    makeTransferTransaction,
    makeOutput,
    makeEd25519Condition,
    signTransaction,
} from 'js-bigchaindb-quickstart';

import { API_PATH } from '../constants/api_urls'
import {
    postTransaction,
    pollStatusAndFetchTransaction,
    requestStatus
} from '../utils/bigchaindb_utils';


console.log("API_PATH:", API_PATH);
// create some identities with a private and a public key
const alice = new Ed25519Keypair();
const bob = new Ed25519Keypair();

let txTransferBobSigned;
// create an asset ie something you want to track or transfer
// think in terms of tokens, versioned data, currency, cars, energy,
// houses, land, diamonds, bikes, permission, personal info, vouchers, ...
const txCreateAliceSimple = makeCreateTransaction(
    {assetMessage: 'I will stick to every future transfer transaction'},
    {metaDataMessage: 'I am specific to this create transaction'},
    [makeOutput(makeEd25519Condition(alice.publicKey))],
    alice.publicKey
);
// sign/fulfill the transaction
const txCreateAliceSimpleSigned = signTransaction(txCreateAliceSimple, alice.privateKey);

// send it off to bigchaindb
console.log('Posting signed transaction: ', txCreateAliceSimpleSigned);
postTransaction(txCreateAliceSimpleSigned)
    .then((res) => {
        console.log('Response from BDB server', res);
        // request the status
        requestStatus(txCreateAliceSimpleSigned.id)
            .then((res) => console.log('Transaction status:', res.status));
        // poll the status every 0.5 seconds
        return pollStatusAndFetchTransaction(txCreateAliceSimpleSigned)
    })
    .then((res) => {
        // the asset can now be transfered
        // every transaction on this asset will carry the asset_id
        // and can be queried by asset_id
        // transfer the asset to bob
        // note that you can append metadata to each transaction
        // these can represent state updates such as:
        // per-transaction-messages, measured values, data stream values
        // incremental update, git commit, ...
        const txTransferBob = makeTransferTransaction(
            txCreateAliceSimpleSigned,
            {metaDataMessage: 'I am specific to this transfer transaction'},
            [makeOutput(makeEd25519Condition(bob.publicKey))], 0);
        // sign with alice's private key
        txTransferBobSigned = signTransaction(txTransferBob, alice.privateKey);
        console.log('d', txTransferBobSigned)
        console.log('Posting signed transaction: ', txTransferBobSigned);
        // post and poll status
        return postTransaction(txTransferBobSigned)
    })
    .then((res) => {
        console.log('Response from BDB server:', res);
        return pollStatusAndFetchTransaction(txTransferBobSigned);
    })

File: examples/client/tutorials/transactions/simple_transactions.py
from bigchaindb_driver import (
    BigchainDB,
)
from bigchaindb_driver.crypto import generate_keypair

from client.tutorials.constants.application_constants import BDB_SERVER_URL
from client.tutorials.utils.bigchaindb_utils import (
    poll_status_and_fetch_transaction,
    prepare_transfer_ed25519_simple,
    sign_ed25519
)


bdb = BigchainDB(BDB_SERVER_URL)

print(bdb.info())

alice = generate_keypair()
bob = generate_keypair()
carly = generate_keypair()

tx_create_alice_simple = bdb.transactions.prepare(
    operation='CREATE',
    signers=alice.public_key,
    asset={'data':
               {'asset_message': 'I will stick to every future transfer transaction'}
           },
    metadata={'metadata_message': 'I am specific to this create transaction'}
)
tx_create_alice_simple_signed = bdb.transactions.fulfill(
    tx_create_alice_simple, private_keys=alice.private_key)

asset_id = tx_create_alice_simple_signed['id']

print('Posting signed transaction{}'.format(tx_create_alice_simple_signed))
res = bdb.transactions.send(tx_create_alice_simple_signed)
poll_status_and_fetch_transaction(tx_create_alice_simple_signed['id'], driver=bdb)

print('Posting signed transaction{}'.format(tx_create_alice_simple_signed))
bdb.transactions.send(tx_create_alice_simple_signed)
poll_status_and_fetch_transaction(tx_create_alice_simple_signed['id'], driver=bdb)

tx_transfer_bob = prepare_transfer_ed25519_simple(
    transaction=tx_create_alice_simple_signed,
    receiver=bob.public_key,
    metadata={'metadata_message': 'I am specific to this transfer transaction'})
tx_transfer_bob_signed = sign_ed25519(tx_transfer_bob, alice.private_key)

print('Posting signed transaction{}'.format(tx_transfer_bob_signed))
bdb.transactions.send(tx_transfer_bob_signed)
poll_status_and_fetch_transaction(tx_transfer_bob_signed['id'], driver=bdb)

res = bdb.transactions.get(asset_id=asset_id)
print('Retrieve list of transactions with asset_id {}: {}'.format(asset_id, len(res)))
res = bdb.outputs.get(public_key=alice.public_key, unspent=True)
print('Retrieve list of unspent outputs with public_key {}: {}'.format(alice.public_key, len(res)))
res = bdb.outputs.get(public_key=bob.public_key)
print('Retrieve list of unspent outputs with public_key {}: {}'.format(bob.public_key, len(res)))
res = bdb.outputs.get(public_key=carly.public_key, unspent=True)
print('Retrieve list of unspent outputs with public_key {}: {}'.format(carly.public_key, len(res)))
